### PR TITLE
[DataGrid] Make GridToolbarXXX props overridable

### LINKED ISSUE #2083: [DataGrid] GridToolbarFilterButton props spread doesn't follow other Toolbar components
<GridToolbarColumnsButton color="inherit" /> render a "inherit" button, <GridToolbarFilterButton color="inherit" /> render a "primary" button.

- [x ] The issue is present in the latest release.
- [x ] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯
GridToolbarColumnsButton props spreading comes before override, not after.

## Expected Behavior 🤔
Props spreading should come after.

## Your Environment 🌎

<details>
  <summary>System info</summary>

System:
    OS: Windows 10 10.0.19043
  Binaries:
    Node: 14.17.2 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 6.14.13 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: Not Found
    Edge: Spartan (44.19041.1023.0), Chromium (91.0.864.64)
  npmPackages:
    @emotion/react: ^11.4.0 => 11.4.0
    @emotion/styled: ^11.3.0 => 11.3.0
    @material-ui/core: ^5.0.0-beta.0 => 5.0.0-beta.0
    @material-ui/data-grid: ^4.0.0-alpha.33 => 4.0.0-alpha.33
    @material-ui/icons: ^5.0.0-beta.0 => 5.0.0-beta.0
    @material-ui/lab: ^5.0.0-alpha.38 => 5.0.0-alpha.38
    @material-ui/private-theming:  5.0.0-alpha.36
    @material-ui/styled-engine:  5.0.0-alpha.34
    @material-ui/styles: ^5.0.0-beta.0 => 5.0.0-beta.0
    @material-ui/system:  5.0.0-alpha.38
    @material-ui/types:  6.0.1
    @material-ui/unstyled:  5.0.0-alpha.39
    @material-ui/utils:  5.0.0-alpha.36
    @types/react: ^17.0.13 => 17.0.13
    react: ^17.0.2 => 17.0.2
    react-dom: ^17.0.2 => 17.0.2
    typescript: ^4.3.5 => 4.3.5
</details>

### PR DESCRIPTION
**Pull request by tifosiblack**
Fix #2083

### CODE REVIEW

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx [hunk 1]
```diff
@@ -95,7 +95,6 @@ export const GridToolbarFilterButton = React.forwardRef<HTMLButtonElement, Butto
       <Tooltip title={tooltipContentNode} enterDelay={1000}>
         <Button
           ref={ref}
-          {...props}
           onClick={toggleFilter}
```
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:98]: I don't think we want to be able to totally skip `toggleFilter`
It should be after the props spread.
And not directly related to this PR, the `onClick` should either be removed from the typing of the component or called in `toggleFilter`. Right now the typing allows to pass a `onClick` but does nothing with it.

Comment (tifosiblack) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:98]: I guess I can exclude it from ButtonProps

Comment (tifosiblack) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:98]: I can do the same for other toolbar component

**Review by oliviertassinari — DISMISSED** [review]
The convention is for the props to be spread on the root element and last. So for instance on the tooltip, not the button. Note that the extra props of the tooltip goes to the button but we probably don't want to talk about it a componentsProps.button would do it.

**Comment by tifosiblack** [comment]
> The convention is for the props to be spread on the root element and last. So for instance on the tooltip, not the button. Note that the extra props of the tooltip goes to the button but we probably don't want to talk about it a componentsProps.button would do it.

title is required for tooltip, should I use `Omit<TooltipProps, 'title'>` or use `Partial<TooltipProps>` ?

**Comment by oliviertassinari** [comment]
Omit sounds great

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx (hunk 1 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx]: > I don't think we want to be able to totally skip toggleFilter

@flaviendelangle Agree, only with  #1403.

**Comment by oliviertassinari** [comment]
Best to review with:

Capture d’écran 2021-07-07 à 01 24 00

**Comment by tifosiblack** [comment]
> Best to review with:
> 
> Capture d’écran 2021-07-07 à 01 24 00

Indeed, I used prettier formatter, but it changed a lot the file whitespaces...

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx [hunk 2]
```diff
@@ -30,85 +30,96 @@ const useStyles = makeStyles(
...
+        ref={ref}
+        size="small"
+        color="primary"
+        aria-label={apiRef!.current.getLocaleText('toolbarFiltersLabel')}
+        startIcon={
+          <Badge badgeContent={counter} color="primary">
+            <OpenFilterButtonIcon />
+          </Badge>
+        }
+        onClick={toggleFilter}
```
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: Either call `componentProps.button?.onClick?.()` in `toggleFilter` or omit `toggleFilter` from the `componentProps.button` interface.

@oliviertassinari do we want to allow the override of any prop ?
For instance here, the user could override `startIcon` which is already customizable with the `OpenFilterButtonIcon` (but he currenctly can't remove the badge).

**Review by flaviendelangle — CHANGES_REQUESTED** [review]

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx (hunk 2 above)
Comment (tifosiblack) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: Not an "easy" change after all ;)
IMO we should be able to change completely the appearance of the button as long as the behavior is kept.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: I'm also in favor of letting the user change the  `startIcon`, and it's a modification the user could do by mistake.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: > do we want to allow the override of any prop ?

@flaviendelangle Yes, I think so. Developers should be able to override `startIcon`.

The case of the event handlers is interesting. We don't have a policy around it yet. In the core, If I look at `SliderUnstyled.js`, `componentProps` overrides the event handlers. If I look at `SwitchUnstyled`, the event handlers of `componentProps` are merged.

cc @michaldudak and @mnajdova. Maybe we should merge?

Comment (michaldudak) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: We should definitively have consistency in this matter across the codebase. My preference for event handlers and classNames is to merge. I suppose it'll create less WTF moments among developers 😉

**Review by flaviendelangle — APPROVED** [review]

File: packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx (hunk 2 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: @flaviendelangle I have noticed that you merged without merging the onClick handlers. Was it on purpose? Is this something we should handle in a follow-up?

Comment (tifosiblack) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: I didn't see you already merged, I may add another PR to deal with event merging.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: > I may add another PR to deal with event merging.

@tifosiblack Flavien is off until Monday, another PR he can review by the time he comes back would be awesome 👌

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/components/toolbar/GridToolbarFilterButton.tsx:118]: Definitely an error on my side

**Pull request merged**